### PR TITLE
BED-4358: Add Complexity Measurement For Graph Mutations

### PR DESCRIPTION
**Pull request by superlinkx**
## Description

This PR adds our initial complexity weights for different mutation statements during query analysis. As a byproduct, it also properly categorizes queries with no constraints (e.g. `match (n) return n`) as beyond the max weight allowed, so that when cypher complexity guardrails are enabled, these queries will be blocked appropriately. A decent number of our test cases needed to be updated to handle this change, and their complexities changed in response too.

Additionally, we fixed an identified bug with our multipart query parsing that caused WITH clauses to break (e.g. `WITH 'sid' AS sid MATCH (u:User {objectid: sid}) RETURN u`).

## Motivation and Context

BED-4358

## How Has This Been Tested?

Added additional tests for complexity and parsing
Updated a few breaking tests as appropriate
Spot checking by running some cypher we were concerned with

## Types of changes

-   [x] Chore (a change that does not modify the application functionality)
-   [x] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: packages/go/cypher/analyzer/measure.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+func (s *ComplexityMeasure) onQuantifier(_ *model.WalkStack, _ *model.Quantifier) error {
+	// Quantifier expressions may increase the size of an inline projection to apply its contained filter and should
+	// be weighted
+	s.Weight += weight1
+	return nil
+}
+
+func (s *ComplexityMeasure) onRelationshipPattern(_ *model.WalkStack, node *model.RelationshipPattern) error {
```
Comment (urangel) [inline packages/go/cypher/analyzer/measure.go:199]: I realize most of this code is just being moved from the analyzer.go file but I was wondering why we name the `*model.RelationshipPattern` parameter as `node`. Similarly for other param names in these methods. I might expect it to be named `relationshipPattern` like in `liftRelationshipPatternCriteria`. 

The reason I was curious was because I see that `match (n) return n` will be gated by setting it to be over the max complexity allowed. I often search for a specific edge in the cypher search, e.g., `match p= ()-[:DCSync]->() return p` which does not have any "node" constraints but does have a relationship constraint. Would DCSync then be included in the `ComplexityMeasure`'s `nodeLookupKinds`? Or will this query for returning these DCSync edges be disallowed because it has no "node" lookup kinds?

File: cmd/api/src/api/v2/cypherquery_integration_test.go [hunk 2]
```diff
@@ -169,7 +168,7 @@ func Test_CypherSearch_WithCypherMutationsEnabled(t *testing.T) {
 			assert.True(ok)
 
 			var (
-				query         = "match (w) where w.name = 'harryp' delete w"
+				query         = "match (w: Wizard) where w.name = 'harryp' delete w"
```
Comment (urangel) [inline cmd/api/src/api/v2/cypherquery_integration_test.go:171]: yer a `Label` now Harry!

File: packages/go/cypher/frontend/query.go [hunk 3]
```diff
@@ -363,43 +362,34 @@ func NewMultiPartQueryVisitor() *MultiPartQueryVisitor {
...
+	part := model.NewMultiPartQueryPart()
+	part.AddReadingClause(s.ctx.Exit().(*ReadingClauseVisitor).ReadingClause)
+	s.Query.Parts = append(s.Query.Parts, part)
 }
 
 func (s *MultiPartQueryVisitor) EnterOC_UpdatingClause(ctx *parser.OC_UpdatingClauseContext) {
 	s.ctx.Enter(NewUpdatingClauseVisitor())
 }
 
 func (s *MultiPartQueryVisitor) ExitOC_UpdatingClause(ctx *parser.OC_UpdatingClauseContext) {
```
Comment (urangel) [inline packages/go/cypher/frontend/query.go:378]: What does "OC" stand for?

File: packages/go/cypher/analyzer/measure.go (hunk 1 above)
Comment (superlinkx) [reply packages/go/cypher/analyzer/measure.go:199]: I'm not entirely sure why node was chosen as the name for these parameters, other than maybe because they are nodes in the parser tree? Regardless, we did a decent amount of work to ensure that _any_ constraint would be sufficient to avoid that max complexity call onExit. That includes relationship labels, node labels, property filters, any where clause, or a limit.

File: packages/go/cypher/frontend/query.go (hunk 3 above)
Comment (superlinkx) [reply packages/go/cypher/frontend/query.go:378]: I'm not entirely sure. These are all overrides for functions that are defined by the antlr generated code

File: packages/go/cypher/analyzer/measure.go (hunk 1 above)
Comment (urangel) [reply packages/go/cypher/analyzer/measure.go:199]: Yea nodes in the parser tree was something I was considering. Okay, that gives me peace of mind to know that edge matching like I wrote would still be allowed. Thanks for the explanation!

File: packages/go/cypher/frontend/query.go (hunk 3 above)
Comment (urangel) [reply packages/go/cypher/frontend/query.go:378]: Cool was just wondering since I didn't see it spelled out anywhere at a glance.

**Review by urangel — APPROVED** [review]
Thanks for the new and updated tests!

File: packages/go/cypher/analyzer/measure.go (hunk 1 above)
Comment (superlinkx) [reply packages/go/cypher/analyzer/measure.go:199]: If you look at the JSON tests, you should see a few tests that validate that an edge label is sufficient as well. We tried to add any tests that seemed to be missing to ensure we'd catch any known edge cases

Comment (superlinkx) [reply packages/go/cypher/analyzer/measure.go:199]:
```
{
    "name": "Query for the properties of an edge using keys()",
    "type": "string_match",
    "details": {
        "query": "match ()-[e:EDGE_OF_INTEREST]->() return keys(e)",
        "complexity": 6
    }
},
```
